presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -258,35 +265,48 @@ private void processNewSources()
     {
         checkLockNotHeld("Can not process while holding the driver lock");
 
+        // if the driver is blocked we don't need to continue
+        if (!previouslyReturnedBlockedFuture.isDone()) {
+            return previouslyReturnedBlockedFuture;
+        }
+
         Optional<ListenableFuture<?>> result = tryWithLock(100, TimeUnit.MILLISECONDS, () -> {
             ListenableFuture<?> driverBlockedFuture = processInternal();
-            return wakeUpOnRevokeRequest(driverBlockedFuture);
+            previouslyReturnedBlockedFuture = wakeUpOnRevokeRequest(driverBlockedFuture);
+            return previouslyReturnedBlockedFuture;
         });
         return result.orElse(NOT_BLOCKED);
     }
 
     private ListenableFuture<?> wakeUpOnRevokeRequest(ListenableFuture<?> driverBlockedFuture)
     {
+        verify(previouslyReturnedBlockedFuture.isDone(), "previouslyReturnedBlockedFuture must be done");
```

Comment:
I think this condition is not very important and could be removed. I mean: even if we remove condition in `process` that prevents coming here until `previouslyReturnedBlockedFuture` is completed, would anything wrong happen _in this method_?

Am I overlooking something?

Reply:
This is for ensuring that when we arrive at this method we don't need the previous listener anymore (set with `setMemoryRevokingRequestedListener` below) and we can set the new one. Otherwise, if the previous future is not done and we set the new listener, the old listener may never get called.

The reason I added this is, the driver process method can be called even when the driver is blocked (mostly due to races per our conversation with @dain, but we need more clarification here). If it turns out that this is not needed, I will be happy to get rid of it.

Reply:
Hm.. we're going to return new blocked future. Who would be holding on to the previous one? 
If you choose to keep this `verify`, then please retain some trace of our discussion here in a code comment.

Reply:
I see that the blocked future is passed all the way up to the `TaskExecutor` threads.
I removed the `verify`.